src/hooks/transactions/useTransactionPreparation.ts

```diff
@@ -33,89 +29,31 @@ interface Props {
-  weightInput,
-  isValidCall = true
-}: Props): FeeAndPayload {
+export default function useTransactionPreparation({ input, type, weightInput, isValidCall = true }: Props) {
   const { areApiReady } = useLoadingApi();
-  const laneId = useLaneId();
+
   const {
-    sourceChainDetails: {
-      apiConnection: { api: sourceApi }
-    },
-    targetChainDetails: { chain: targetChain }
+    sourceChainDetails: { chain: sourceChain }
   } = useSourceTarget();
   const { account } = useAccountContext();
-
-  const [payload, setPayload] = useState<null | {}>(null);
   const { call, weight } = useTransactionType({ input, type, weightInput });
 
-  const { dispatchTransaction } = useUpdateTransactionContext();
-  const { estimatedFeeMethodName } = getSubstrateDynamicNames(targetChain);
+  const calculateFee = useEstimateFee();
+  const updatePayload = usePayload();
 
   useEffect(() => {
-    const calculateFee = async () => {
-      // Ignoring custom types missed for TS for now.
-      // Need to apply: https://polkadot.js.org/docs/api/start/typescript.user
-      // @ts-ignore
-      const payloadType = sourceApi.registry.createType('OutboundPayload', payload);
-      // @ts-ignore
-      const messageFeeType = sourceApi.registry.createType('MessageFeeData', {
-        lane_id: laneId,
-        payload: payloadType.toHex()
-      });
-
-      const estimatedFeeCall = await sourceApi.rpc.state.call<Codec>(estimatedFeeMethodName, messageFeeType.toHex());
-
-      // @ts-ignore
-      const estimatedFeeType = sourceApi.registry.createType('Option<Balance>', estimatedFeeCall);
-      const estimatedFee = estimatedFeeType.toString();
-
-      dispatchTransaction(TransactionActionCreators.setEstimateFee(estimatedFee));
+    const asyncCalculateFee = async () => {
+      await calculateFee();
     };
-
-    if (areApiReady && payload) {
-      calculateFee();
+    if (areApiReady) {
+      asyncCalculateFee();
     }
-  }, [
-    areApiReady,
-    dispatchTransaction,
-    estimatedFeeMethodName,
-    laneId,
-    payload,
-    sourceApi.registry,
-    sourceApi.rpc.state,
-    targetChain
-  ]);
+  }, [areApiReady, calculateFee]);
 
   useEffect(() => {
-    if (!(isValidCall && account && call && weight)) {
+    if (!isValidCall) {
       return;
     }
-
-    const payload = {
-      call: compactAddLength(call),
-      origin: {
-        SourceAccount: account.addressRaw
-      },
-      // TODO [#122] This must not be hardcoded.
-      spec_version: 1,
-      weight
-    };
-    // @ts-ignore
-    const payloadType = sourceApi.registry.createType('OutboundPayload', payload);
-    logger.info(`OutboundPayload: ${JSON.stringify(payload)}`);
-    logger.info(`OutboundPayload.toHex(): ${payloadType.toHex()}`);
-    setPayload(payload);
-  }, [account, call, isValidCall, type, weight, sourceApi.registry]);
-
-  return {
-    payload
-  };
+    updatePayload(call, weight);
+  }, [account, call, isValidCall, type, weight, sourceChain, updatePayload]);
```

Comment:
This is extremely confusing and another source of bugs. Why are all of these deps here? If they are really needed that's something seriously worrisome, but I hope they are just forgotten during refactoring - again if that's the case I really think this kind of things should be caught during self-review, which I kindly ask you to do.

```suggestion
  }, [call, isValidCall, weight, updatePayload]);
```

Reply:
Code updated. Looks better now.